XWIKI-7716: Use .xform style standard in the User Profile

Usamaahmad751 opened the pull request:
Issue XWIKI-7716 has been solved.
Issue-Link: https://jira.xwiki.org/browse/XWIKI-7716
Replacing h1 tags with h2 tags. As there should be only one h1 tag in the document.
Added a title to a link `<a>`  tag.
**Result:** 
![Issue-7716-solved](https://user-images.githubusercontent.com/40354433/75925637-d4fbf000-5e8a-11ea-901c-0d042444b04e.PNG)
**Inspect**
![Issue-7716-solved-inspect](https://user-images.githubusercontent.com/40354433/75925825-37ed8700-5e8b-11ea-8d87-8e59dfa80866.PNG)

surli commented:
Hi @Usamaahmad751 thanks for your contrib. It looks good to me, but I don't know well our styling standard for CSS :) 

> According to .xform style standard, there should be .xHelp class in the tag.

Could you point me out where you find this? I had a quick look but didn't find it.

Some general observation:
  - be careful, you forgot to assign XWIKI-7716 to yourself, it should be the first thing to do when you start working on an issue, else someone else could propose a solution and it's time lost
  - if possible provide the link to the issue on the first comment of the PR, it makes reviewer work easier :)

Usamaahmad751 commented:
Hi, surli! Thanks for reviewing my PR. I added the .xHelp class to <a> tag according to the standards defined in the table. Please visit this link and slide down to a table I am referring to.
https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/FrontendResources/VerticalForms

If you think that does not need to be added please let me know I will make changes accordingly. Thank You!

surli commented:
> I added the .xHelp class to a tag according to the standards defined in the table. Please visit this link and slide down to a table I am referring to.
xwiki.org/xwiki/bin/view/Documentation/DevGuide/FrontendResources/VerticalForms

Thanks for pointing out the doc. Actually this `.xHelp` class is normally used to add an "help" icon on a form, you have a demo there: 
https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/FrontendResources/VerticalForms#H3.Help

Here it's really not what we want. Now I think that we should add a title to the link, to help people using accessibility tools.

Usamaahmad751 commented:
Oh! Got it.  Yes, that will be great to add a title to the link. Let me make the changes. Thank You!

Usamaahmad751 commented:
Hi @surli! please review the updated file now. I removed the .xHelp class and added a title to a link tag. Let me know if I made any mistakes. Thank You!

surli commented:
That part (replace H1 to H2 and add the title) looks good to me.
What about the other work mentioned in the issue:
> and labels have a .label class associated with it. We need either to remove them or revise the standard and skin's styling.

Usamaahmad751 commented:
`.label` class is added to a container of the `<label>` that is `<dt>`. 
Like:
 `<dt class="label">`
`<label>Last Name</label>`
`</dt>`
but not directly to `<label>`. So I though it's not associated directly with `<label>`. That;s why I didn't make change to `.label` class. 
But If it's required to remove `.label` class from the container. I will make a change.

Usamaahmad751 commented:
Hi@surli! Now Please have a look at it. I think now the issue has been resolved. If there is any mistake please let me know. Thanks

surli commented:
If you could just repost a new screenshot to be sure removing the label won't have any impact, else I'm all good.

Usamaahmad751 commented:
Final Result: 
![Result](https://user-images.githubusercontent.com/40354433/76013657-1d261b80-5f3a-11ea-9030-4cfd2a2ba08d.PNG)

**Inspect:** 

` class=".label" ` is removed from container `<dt>`
![Inspect](https://user-images.githubusercontent.com/40354433/76013750-4777d900-5f3a-11ea-9eb1-142ab02d2f18.PNG)

Usamaahmad751 commented:
@surli I have added a CSS style to make the heading `<h2>` bold and it's font-size little bigger. By applying that style section-heading becomes more prominent. As It can be seen in screen-shots.
Should I also send a pull request to add its CSS style?

surli commented:
> Should I also send a pull request to add its CSS style?

Could you post the 2 screenshots, with an without your CSS changes (but in both case with the changes made on the html). 
IMO XWIKI-7716 was only on changes to the html, but we should obtain a similar result in terms of UI at the end. Now, I'm not entirely sure where a new CSS should be added for this part. @mflorea wdyt?

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -89,7 +89,7 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
     &lt;div class='userInfo'&gt;
     #if($xcontext.action == 'view' &amp;&amp; $hasEdit)
-      &lt;div class='editProfileCategory'&gt;&lt;a href="$doc.getURL('edit', 'editor=inline&amp;amp;category=profile')"&gt;&lt;span class='hidden'&gt;$services.localization.render('platform.core.profile.category.profile.edit')&lt;/span&gt;&lt;/a&gt;&lt;/div&gt;
+      &lt;div class='editProfileCategory'&gt;&lt;a title="$services.localization.render('platform.core.profile.category.profile.edit')" href="$doc.getURL('edit', 'editor=inline&amp;amp;category=profile')"&gt;&lt;span class='hidden'&gt;$services.localization.render('platform.core.profile.category.profile.edit')&lt;/span&gt;&lt;/a&gt;&lt;/div&gt;
```
This line needs to be formatted to not exceed 120 character limit. The title attribute value needs to be XML escaped using ``$escapetool.xml(...)``. The edit link label should probably use ``sr-only`` instead of ``hidden`` (so that screen readers read it).

Usamaahmad751 replied:
@mflorea Is It the right approach to escape xml and limiting the characters to 120?

`#set ( $editProfileCategory = $escapetool.xml($services.localization.render('platform.core.profile.category.profile.edit')))`
 
**Then** 
`title="$editProfileCategory"`

mflorea replied:
That is an option, but not the only one (and maybe not the best). Start by formatting the HTML as it should be.

mflorea commented:
@Usamaahmad751 can you post a screenshot without the custom CSS styles you added?

Usamaahmad751 commented:
> @Usamaahmad751 can you post a screenshot without the custom CSS styles you added?

**Before styling:**

![Before-CSS-h2](https://user-images.githubusercontent.com/40354433/76077085-dbd94e80-5fc1-11ea-8e95-7b7db4c6ffe6.PNG)

**After styling:**

![Result](https://user-images.githubusercontent.com/40354433/76077137-f8758680-5fc1-11ea-960a-3b40b761658f.PNG)

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -145,9 +145,9 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
+        &lt;h2&gt;$services.localization.render('platform.core.profile.section.activity')&lt;/h2&gt;
       #else
-        &lt;h1&gt;$services.localization.render('platform.core.profile.section.activityof', [$xwiki.getUserName($doc.fullName, false)])&lt;/h1&gt;
+        &lt;h2&gt;$services.localization.render('platform.core.profile.section.activityof', [$xwiki.getUserName($doc.fullName, false)])&lt;/h2&gt;
```
This line is also too long, and since you changed it, it would be good to format it properly.

mflorea commented:
We need the CSS styles indeed. Should be added to this PR.

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -89,7 +89,12 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
     #if($xcontext.action == 'view' &amp;&amp; $hasEdit)
-      &lt;div class='editProfileCategory'&gt;&lt;a href="$doc.getURL('edit', 'editor=inline&amp;amp;category=profile')"&gt;&lt;span class='hidden'&gt;$services.localization.render('platform.core.profile.category.profile.edit')&lt;/span&gt;&lt;/a&gt;&lt;/div&gt;
+      &lt;div class='editProfileCategory'&gt;
+       &lt;a title="$escapetool.xml($services.localization.render('platform.core.profile.category.profile.edit'))" 
```
We indent HTML using 2 spaces.

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -89,7 +89,12 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
-      &lt;div class='editProfileCategory'&gt;&lt;a href="$doc.getURL('edit', 'editor=inline&amp;amp;category=profile')"&gt;&lt;span class='hidden'&gt;$services.localization.render('platform.core.profile.category.profile.edit')&lt;/span&gt;&lt;/a&gt;&lt;/div&gt;
+      &lt;div class='editProfileCategory'&gt;
+       &lt;a title="$escapetool.xml($services.localization.render('platform.core.profile.category.profile.edit'))" 
+        href="$doc.getURL('edit', 'editor=inline&amp;amp;category=profile')"&gt;
```
This line needs to be indented even more in order to be clear that it's not on the same level as the child ``span``.

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -89,7 +89,12 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
+      &lt;div class='editProfileCategory'&gt;
+       &lt;a title="$escapetool.xml($services.localization.render('platform.core.profile.category.profile.edit'))" 
+        href="$doc.getURL('edit', 'editor=inline&amp;amp;category=profile')"&gt;
+        &lt;span class='sr-only'&gt;$services.localization.render('platform.core.profile.category.profile.edit')&lt;/span&gt;
```
The text within the ``span`` element needs to be XML encoded also (like you did above).

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -145,9 +150,12 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
-        &lt;h1&gt;$services.localization.render('platform.core.profile.section.activityof', [$xwiki.getUserName($doc.fullName, false)])&lt;/h1&gt;
+        &lt;h2&gt;
+         $services.localization.render('platform.core.profile.section.activityof',
+         [$xwiki.getUserName($doc.fullName, false)])
```
This line needs to be indented one level more, since it's a fragment of the previous line of code.

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserSheet.xml:
```diff
@@ -585,7 +585,12 @@ div.userInfo, div.userPreferences, div.watchlistManagement, div.userDashboard {
-
+        
+.userInfo h2 {
+  font-size: 16px;
```
Using absolute values is not a good practice in general, and especially in the case of font size. Moreover, pixels is not a great unit to specify the font size. Try a relative value / unit. How does it look for instance with:

```
font-size: larger;
font-weight: bolder;
```

Usamaahmad751 replied:
I replace the `font-size` value from `px` to `large`. On relative value `large`. It gives the same result as above in the screenshot.

mflorea replied:
``large`` is not a relative value. ``larger`` is. See https://developer.mozilla.org/en-US/docs/Web/CSS/font-size#Syntax .

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -145,9 +150,12 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
       &lt;div class='userRecentChanges'&gt;
       #if ($xcontext.user == $doc.fullName)
-        &lt;h1&gt;$services.localization.render('platform.core.profile.section.activity')&lt;/h1&gt;
+        &lt;h2&gt;$services.localization.render('platform.core.profile.section.activity')&lt;/h2&gt;
```
Since you're modifying this line, would be good to also escape the XML from the heading text using ``$escapetool.xml()``.

mflorea commented on xwiki-platform-core/xwiki-platform-user/xwiki-platform-user-profile/xwiki-platform-user-profile-ui/src/main/resources/XWiki/XWikiUserProfileSheet.xml:
```diff
@@ -89,7 +89,12 @@ $services.localization.render('platform.core.profile.category.profile.disabled')
+        &lt;a title="$escapetool.xml($services.localization.render('platform.core.profile.category.profile.edit'))" 
+              href="$doc.getURL('edit', 'editor=inline&amp;amp;category=profile')"&gt;
+        &lt;span class='sr-only'&gt;$escapetool.xml($services.localization.render('platform.core.profile.category.profile.edit'))&lt;/span&gt;
+       &lt;/a&gt;
```
Bad formatting. The closing ``</a>`` should be on the same level as the starting ``<a>`` and the child ``<span>`` should be indented..

mflorea commented:
Thanks for your contribution.
